app/src/androidTest/kotlin/com/google/samples/apps/nowinandroid/ui/interests2pane/InterestsListDetailScreenTest.kt

```diff
@@ -0,0 +1,231 @@
+import androidx.compose.material3.adaptive.Posture
+import androidx.compose.material3.adaptive.WindowAdaptiveInfo
+import androidx.compose.ui.test.DeviceConfigurationOverride
+import androidx.compose.ui.test.ForcedSize
+import androidx.compose.ui.test.assertIsDisplayed
+import androidx.compose.ui.test.assertIsNotDisplayed
+import androidx.compose.ui.test.junit4.createAndroidComposeRule
+import androidx.compose.ui.test.onNodeWithTag
+import androidx.compose.ui.test.onNodeWithText
+import androidx.compose.ui.test.performClick
+import androidx.compose.ui.unit.DpSize
+import androidx.compose.ui.unit.dp
+import androidx.test.espresso.Espresso
+import androidx.window.core.layout.WindowSizeClass
+import com.google.samples.apps.nowinandroid.core.data.repository.TopicsRepository
+import com.google.samples.apps.nowinandroid.core.designsystem.theme.NiaTheme
+import com.google.samples.apps.nowinandroid.core.model.data.Topic
+import com.google.samples.apps.nowinandroid.ui.stringResource
+import com.google.samples.apps.nowinandroid.uitesthiltmanifest.HiltComponentActivity
+import dagger.hilt.android.testing.BindValue
+import dagger.hilt.android.testing.HiltAndroidRule
+import dagger.hilt.android.testing.HiltAndroidTest
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.runBlocking
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TemporaryFolder
+import javax.inject.Inject
+import kotlin.test.assertTrue
+import com.google.samples.apps.nowinandroid.feature.topic.R as FeatureTopicR
+
+@HiltAndroidTest
+class InterestsListDetailScreenTest {
+    @get:Rule(order = 0)
+    val hiltRule = HiltAndroidRule(this)
+
+    @BindValue
+    @get:Rule(order = 1)
+    val tmpFolder: TemporaryFolder = TemporaryFolder.builder().assureDeletion().build()
+
+    @get:Rule(order = 2)
+    val composeTestRule = createAndroidComposeRule<HiltComponentActivity>()
+
+    @Inject
+    lateinit var topicsRepository: TopicsRepository
+
+    // The strings used for matching in these tests.
+    private val placeholderText by composeTestRule.stringResource(FeatureTopicR.string.feature_topic_select_an_interest)
+    private val listPaneTag = "interests:topics"
+
+    private val Topic.testTag
+        get() = "topic:${this.id}"
+
+    // Overrides for device sizes.
+    private enum class TestDeviceConfig(widthDp: Float, heightDp: Float) {
+        Compact(412f, 915f),
+        Expanded(1200f, 840f);
+
+        val sizeOverride = DeviceConfigurationOverride.ForcedSize(DpSize(widthDp.dp, heightDp.dp))
+        val adaptiveInfo = WindowAdaptiveInfo(
+            windowSizeClass = WindowSizeClass.compute(widthDp, heightDp),
+            windowPosture = Posture()
+        )
+    }
+
+    @Before
+    fun setup() {
+        hiltRule.inject()
+        composeTestRule.activityRule
+    }
+
+    @Test
+    fun expandedWidth_initialState_showsTwoPanesWithPlaceholder() {
+        composeTestRule.apply {
```

Comment:
I feel like these patterns of `composeTestRule.apply` followed by `setContent` could be replaced with a more simple `composeTestRule.setContent` and also to reduce indentation.


Or even simpler, if all these tests are actually doing more or less the same thing, to extract the setup phase into a private function:

```kotlin
private fun setup(
    config: TestDeviceConfig,
    then: AndroidComposeTestRule<*, *>.() -> Unit,
) = with(composeTestRule) {
    setContent {
        DeviceConfigurationOverride(override = config.sizeOverride) {
            NiaTheme {
                InterestsListDetailScreen(windowAdaptiveInfo = config.adaptiveInfo)
            }
        }
    }
    then()
}
```

to be called like this in the tests themselves:

```kotlin
setup(TestDeviceConfig.Compact) {
    onNodeWithTag(...).assertIsDisplayed()
    ...
}
```

Reply:
Agreed that we could probably use `= composeTestRule.apply {` to save a level of indention which might be nice.

I'm not opposed to a setup function if it is the same, but since the `expandedWidth_backPressFromTopicDetail_leavesInterests` test has a slightly different setup with the `BackHandler` to test that behavior, trying to have that configurability in a setup would probably just be more complicated.

I personally don't mind some duplication in tests for the ease of reading and simplicitly. If a test fails, each setup, helper, or other utility method adds an additional place where you have to dive deeper and understand what's going on in the test itself, before figuring out what the actual code is doing.

Reply:
Yeah, it did look a little odd having one test didn't use the helper function. I think the UI setup is small enough that I don't mind repeating it. We can revisit it later if necessary.